lib/HTFeed/JobMetrics.pm

```diff
@@ -0,0 +1,352 @@
+
+=pod
+
+HTFeed::JobMetrics is a class that can be used to gather metrics on
+the various stages of an ingest job, such as number of items downloaded,
+number of bytes downloaded and amount of time spent downloading
+(as well as the other job stages, not just download).
+
+Command line options:
+
+=item --help / --usage
+
+Display the POD snippets in this file as a help document.
+
+=cut
+
+# If called from terminal:
+if (!caller) {
+    GetOptions(
+	# these are all booleans
+	'clear'    => \my $clear,
+	'help'     => \my $help,
+	'list'     => \my $list,
+	'location' => \my $location,
+	'pretty'   => \my $pretty,
+	'usage'    => \my $usage,
+	# each operation take a metric and/or a value:
+	# add(metric, value), inc(metric), match(value), observe(metric, value)
+	'operation=s' => \my $operation,
+	'metric=s'    => \my $metric,
+	'value=s'     => \my $value
+    );
+
+    if ($help or $usage) {
+	pod2usage(2);
+    }
+
+    my $jm = HTFeed::JobMetrics->new();
+    my %valid_operations = (
+	add   => sub { $jm->add($metric, $value) },
+	inc   => sub { $jm->inc($metric) },
+	match => sub {
+	    my $matches = $jm->match($value);
+	    print join("\n", @$matches) . "\n";
+	},
+	observe => sub { $jm->observe($metric, $value) }
+    );
+
+    # Check the boolean args and act if true.
+    if ($clear) {
+	print "Clearing data...\n";
+	$jm->clear;
+	print "Data cleared.\n";
+    }
+    if ($list) {
+	print "# List of metrics:\n";
+	print join("\n", @{$jm->list_metrics}) . "\n";
+    }
+    if ($location) {
+	print "# HTFeed::JobMetrics data stored in:\n";
+	print $jm->loc . "\n";
+    }
+    if ($pretty) {
+	print $jm->pretty . "\n";
+    }
+
+    # Now check the operation arg
+    if (defined $operation) {
+	if (exists $valid_operations{$operation}) {
+	    print "running operation $operation ...\n";
+	    $valid_operations{$operation}();
+	} else {
+	    print "Invalid operation, exiting.\n";
+	    exit(1);
+	}
+    } else {
+	print "No operation defined, exiting.\n";
+	exit(0);
+    }
+}
+# end commandline stuff
+
+sub new {
+    my $class = shift;
+    my $self  = {};
+
+    # TODO: Need to determine best place to store metrics
+    # pod's /tmp? /htprep? some other location?
+    $self->{file} = "/tmp/HTFeed::JobMetrics::Data";
+    if (defined $ENV{'HTFEED_JOBMETRICS_DATA_DIR'}) {
+	$self->{file} = $ENV{'HTFEED_JOBMETRICS_DATA_DIR'};
+    }
+    $self->{prom} = Prometheus::Tiny::Shared->new(
+	filename => $self->{file}
+    );
+
+    bless($self, $class);
+    $self->_setup_metrics();
+
+    return $self;
+}
+
+=item --loc
+
+Show where the data is stored.
+
+=cut
+
+sub loc {
+    my $self = shift;
+    $self->{file};
+}
+
+=item --list
+
+Show names of all known metrics.
+
+=cut
+
+sub list_metrics {
+    my $self = shift;
+    [sort keys %{$self->{metrics}}];
+}
+
+=item --pretty
+
+Full, somewhat readable, accounting of known metrics and their values.
+
+=cut
+
+sub pretty {
+    my $self = shift;
+    $self->{prom}->format;
```

Comment:
Ah. Yes. I added a sentence to the sub's pod to reflect that.
Maybe it would make sense to name it "export" or "scrape" but what it really is is a pretty-printer.